src/xml2_url.cpp

```diff
@@ -1,156 +1,133 @@
+#include <cpp11.hpp>
+
 #define R_NO_REMAP
 #include <Rinternals.h>
 #undef R_NO_REMAP
 
 #include <libxml/uri.h>
 #include "xml2_utils.h"
 
-// [[export]]
-extern "C" SEXP url_absolute_(SEXP x_sxp, SEXP base_sxp) {
-  R_xlen_t n = Rf_xlength(x_sxp);
-  SEXP out = PROTECT(Rf_allocVector(STRSXP, n));
-
-  if (Rf_xlength(base_sxp) > 1) {
-    Rf_error("Base URL must be length 1");
+const xmlChar* to_xml_chr(cpp11::strings x, const char* arg) {
+  if (x.size() > 1) {
+    cpp11::stop("%s must be a character vector of length 1", arg);
   }
 
-  const xmlChar* base_uri = (xmlChar*) Rf_translateCharUTF8(STRING_ELT(base_sxp, 0));
+  return (xmlChar*) cpp11::as_cpp<const char*>(x);
+}
+
+[[cpp11::register]]
+cpp11::strings url_absolute_(cpp11::strings x_sxp, cpp11::strings base_sxp) {
+  int n = x_sxp.size();
+  cpp11::writable::strings out(n);
+
+  const xmlChar* base_uri = to_xml_chr(base_sxp, "Base URL");
 
   for (int i = 0; i < n; ++i) {
-    const xmlChar* uri = (xmlChar*) Rf_translateCharUTF8(STRING_ELT(x_sxp, i));
-    SET_STRING_ELT(out, i, Xml2String(xmlBuildURI(uri, base_uri)).asRString());
+    const xmlChar* uri = (xmlChar*) Rf_translateCharUTF8(x_sxp[i]);
+    out[i] = Xml2String(xmlBuildURI(uri, base_uri)).asRString();
```

Comment:
I guess the slow down due to cpp11 protection probably doesn't matter so much here

Reply:
Yea if it isn't critical then I'd probably leave as is